### PR TITLE
Email notifications

### PR DESCRIPTION
**Pull request by dolgopolovwork**
### Description of the Change
Notifications - is a standalone service dedicated to email notifications. It listens to Iroha blocks and notifies clients if deposit or withdrawal event occurred. We assume that client email is set by himself and stored in account details. 
We also need an SMTP server in order to make this service functioning. SMTP configurations are hidden from VCS since it holds SMTP credentials. But you can use [mailtrap.io](http://mailtrap.io) services for testing purposes(or just ask me, because I've already got mailtrap credentials).

### CODE REVIEW

File: notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+@Component
+class EmailNotificationService(
+    @Autowired private val smtpService: SMTPService,
+    @Autowired private val d3ClientProvider: D3ClientProvider
+) : NotificationService {
+
+    override fun notifyDeposit(transferNotifyEvent: TransferNotifyEvent): Result<Unit, Exception> {
+        logger.info { "Notify deposit $transferNotifyEvent" }
+        val message =
+            "Hello ${transferNotifyEvent.accountId}. We are happy to notify that ${transferNotifyEvent.amount} ${transferNotifyEvent.assetName} has been deposited to your account."
```
Comment (Mingela) [inline notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt]: Hello **dear**
Are we really happy? 🤣 
I think it is better to ask @LiraLemur (https://github.com/LiraLemur) to help with that

File: notifications/src/main/kotlin/notifications/smtp/SMTPService.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            })
+        }.map { session ->
+            val mimeMessage = MimeMessage(session)
+            mimeMessage.setFrom(InternetAddress(from))
+            mimeMessage.setRecipients(
+                Message.RecipientType.TO, InternetAddress.parse(to)
+            )
+            mimeMessage.subject = subject
+            val mimeBodyPart = MimeBodyPart()
+            mimeBodyPart.setContent(message, "text/html")
```
Comment (Mingela) [inline notifications/src/main/kotlin/notifications/smtp/SMTPService.kt]: I know this is the initial version but for future I suggest to use email template, for example, [FreeMarker](https://freemarker.apache.org) (flexible CSS styles support)

File: notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt (hunk 1 above)
Comment (LiraLemur) [reply notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt]: Is is the final state? I think something luke "Dear _. N tokens were deposited to your account." would be sufficient :) Though I am not sure about the context

Comment (dolgopolovwork) [reply notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt]: Oh, thank you @LiraLemur . Can you please also take a look at withdrawal message?

File: notary-commons/src/main/kotlin/provider/D3ClientProvider.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ */
+class D3ClientProvider(private val notaryQueryAPI: QueryAPI) {
+
+    /**
+     * Returns D3 client by its id
+     * @param accountId - account id of client
+     * @return D3 client
+     */
+    fun getClient(accountId: String): Result<D3Client, Exception> {
+        // Assuming that client sets details himself
```
Comment (Mingela) [inline notary-commons/src/main/kotlin/provider/D3ClientProvider.kt]: hmm it's needed to be decided
probably `registration_service` could also take care of it

File: notifications-integration-test/src/integration-test/kotlin/notifications/NotificationsIntegrationTest.kt [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            D3_CLIENT_EMAIL_KEY,
+            "user@d3.com"
+        ).failure { ex -> throw ex }
+
+        // Setting dest client email
+        ModelUtil.setAccountDetail(
+            environment.destClientConsumer,
+            environment.destClientId,
+            D3_CLIENT_EMAIL_KEY,
+            "user@d3.com"
```
Comment (Mingela) [inline notifications-integration-test/src/integration-test/kotlin/notifications/NotificationsIntegrationTest.kt]: just to notice you are using the same email for both clients

File: notifications/src/main/kotlin/notifications/init/NotificationInitialization.kt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                }, { ex ->
+                    logger.error("Error on Iroha subscribe", ex)
+                    onIrohaChainFailure()
+                })
+        }
+    }
+
+    // Checks if deposit event
+    private fun isDeposit(transferAsset: Commands.TransferAsset): Boolean {
+        return transferAsset.srcAccountId.endsWith("@$NOTARY_DOMAIN")
```
Comment (Mingela) [inline notifications/src/main/kotlin/notifications/init/NotificationInitialization.kt]: Do we need to include `@`? Are you assuming we are going to have a domain like `@blablablanotary`? In that case it is better to assign `@notary` to `NOTARY_DOMAIN` unless it is used somewhere else separately

File: notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import com.github.kittinunf.result.Result
+import com.github.kittinunf.result.flatMap
+import mu.KLogging
+import notifications.smtp.SMTPService
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import provider.D3ClientProvider
+
+private const val NOTIFICATION_EMAIL = "noreply@d3.com"
+const val D3_WITHDRAWAL_EMAIL_SUBJECT = "D3 withdrawal"
```
Comment (Mingela) [inline notifications/src/main/kotlin/notifications/service/EmailNotificationService.kt]: I think it is common practice just to use the same subject name for all notifications. Smth like `D3 notification` should be fine

**Comment by Mingela** [comment]
Also could you provide some sample `smtp.properties` ?

File: notifications/src/main/kotlin/notifications/init/NotificationInitialization.kt (hunk 5 above)
Comment (dolgopolovwork) [reply notifications/src/main/kotlin/notifications/init/NotificationInitialization.kt]: Yes, I think that `@blablanotarty` may happen. `NOTARY_DOMAIN` represents the domain. `notary` is a valid domain, while `@notary` is not.

File: notary-commons/src/main/kotlin/provider/D3ClientProvider.kt (hunk 3 above)
Comment (dolgopolovwork) [reply notary-commons/src/main/kotlin/provider/D3ClientProvider.kt]: Unfortunately, we use different `registration_service` accounts for Bitcoin and Ethereum. I suppose that crypto related details will be set by `registration_service` while common information (such as email) will be set by user himself.

**Pull request merged**